### PR TITLE
Add domain for user profile

### PR DESCRIPTION
**Pull request by HenrikJannsen**

### CODE REVIEW

**Review by rodvar — CHANGES_REQUESTED** [review]
great work @HenrikJannsen 

## Runtime

 - I've tested in android pixel 6 (crashes, but expected from your comments) and in emulator. I'll test in all 3 apps on once this first pass of review gets done
 - can get the emulator to run and communicate with running local bisq2-rest-api
 - can see the profile creation working in the logs and the screen 
 - however, I can't go past the screen when clicking next nothing happens and no error is logged
  - some other errors seen:
  ```
  kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 2: Encountered an unknown key 'error' at path: $`
                                                                                                     Use 'ignoreUnknownKeys = true' in 'Json {}' builder to ignore unknown keys.
                                                                                                     JSON input: {"error":"Could not find a selected user identity"}
```

## Code

I don't wanna be too verbose here, I think when you read my individual comments you'll find the resolution to our discussions.  Please reach out if something is not clear.

Regarding the Ktor addition. I was working on this as well but in a separate new `networking` module in an effort to avoid making the androidModule grasp those dependencies when its not gonna be used. 

I think I'll drop that for now until we get this one to be merged and then work on that refactor later on.

I contribute with you by finishing the lifecycle setup so you can reuse in your PR as well, might also include the solution for the Dispatchers.IO for you
cheers!

**Comment by rodvar** [comment]
@HenrikJannsen done this for you

 - https://github.com/bisq-network/bisq-mobile/pull/64
 
 you'll probably get conflicts with the CreateProfile screen related code on rebase. I humbly suggest that we keep the presenter view defined interface (you'll see why in the code) but it's up to you for this particular case. cheers!

**Comment by HenrikJannsen** [comment]
> * I can't go past the screen when clicking next nothing happens and no error is logged

I commented out the code to go to next screen in the presenter for dev testing. see comments there. i wanted to test several use cases and as the UI is not yet prepared for that i did it with changing the presenter. I will add a commit to undo that. But would be good as next step to make the UI behaving to the intended use case.

1. After loading screen -> check if there is an existing user profile. 
1a. If there is an existing user profile, do not show create user profile screen, but show user profile is some not yet defined way (right top corner in Desktop shows user profile).
1b. If there is no existing user profile, show create profile screen.

Additionally: Show busy animation when generating keypair and when creating the profile. The domain model contains 2 boolean `inProgress` properties for that.

**Comment by HenrikJannsen** [comment]
> Regarding the Ktor addition. I was working on this as well but in a separate new `networking` module in an effort to avoid making the androidModule grasp those dependencies when its not gonna be used.

Ah that is good. But maybe we call it 'client' instead of networking, as networking happens also in the node when all is delegated to the libraries.
Is it basically the code from shared.domain.client.
We should also move the Bisq2 replicated models (annotated with @Serializable) which are only used for client over there. 

Btw. I did not clean up the gradle dependencies, there might be too many and they should be moved to toml file.... also there are those issues with the exclusion of the logging lib... But I guess you have better ideas when it comes to gradle stuff..

File: bisqapps/androidClient/src/androidMain/kotlin/network/bisq/mobile/client/di/AndroidClientModule.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import network.bisq.mobile.domain.client.main.user_profile.ClientUserProfileModel
+import network.bisq.mobile.domain.client.main.user_profile.ClientUserProfileServiceFacade
+import network.bisq.mobile.domain.client.main.user_profile.UserProfileApiGateway
+import network.bisq.mobile.domain.user_profile.UserProfileModel
+import network.bisq.mobile.domain.user_profile.UserProfileServiceFacade
+import org.koin.dsl.module
+
+val androidClientModule = module {
+    single<UserProfileModel> { ClientUserProfileModel() }
+    single { ApiRequestService("10.0.2.2") }
```
[the remark this answers is not in the document]
Comment (HenrikJannsen) [reply bisqapps/androidClient/src/androidMain/kotlin/network/bisq/mobile/client/di/AndroidClientModule.kt:13]: I assume that is not in place yet, right? It also need to be customized when running on a device compared to simulator.

[the remark this answers is not in the document]
Comment (rodvar) [reply bisqapps/androidClient/src/androidMain/kotlin/network/bisq/mobile/client/di/AndroidClientModule.kt:13]: not in place yet - we can do it in the context of this PR if you like or on the next one (with the module refactor I was talking above)

File: bisqapps/androidNode/src/androidMain/kotlin/network/bisq/mobile/android/node/di/AndroidNodeModule.kt [hunk 2]
```diff
@@ -1,23 +1,29 @@
...
-import org.koin.android.ext.koin.androidContext
 import org.koin.dsl.bind
 import org.koin.dsl.module
 
 val androidNodeModule = module {
     // this one is for example properties, will be eliminated soon
     single<NodeGreetingRepository> { NodeGreetingRepository() }
-    // this line showcases both, the posibility to change behaviour of the app by changing one definiton
+
+    single { AndroidApplicationService.Supplier() }
```
[the remark this answers is not in the document]
Comment (HenrikJannsen) [reply bisqapps/androidNode/src/androidMain/kotlin/network/bisq/mobile/android/node/di/AndroidNodeModule.kt:19]: > Shouldn't then the AndroidApplicationService instance be configured here

If we manage to get the constructor arguments `context: Context, userDataDir: Path?` available in KOIN module, we can do that. But those data are only available when MainActivity is ready and we create KOIN earlier. But if you have any idea, feel free to change.

> the supplier injected into that service to provide for all the rest of the instances
If you mean how I interpret it, that's the case.
```
single<UserProfileServiceFacade> { NodeUserProfileServiceFacade(get(), get()) }
 
class NodeUserProfileServiceFacade(
    override val model: UserProfileModel,
    private val applicationServiceSupplier: AndroidApplicationService.Supplier
)
```

File: bisqapps/shared/domain/src/commonMain/kotlin/network/bisq/mobile/client/user_profile/ClientUserProfileServiceFacade.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+import kotlin.random.Random
+
+class ClientUserProfileServiceFacade(
+    override val model: UserProfileModel,
+    private val apiGateway: UserProfileApiGateway
+) :
+    UserProfileServiceFacade {
+    private val log = Logger.withTag("IosClientUserProfileController")
+
+    // TODO Dispatchers.IO is not supported on iOS. Either customize or find whats on iOS appropriate.
```
[the remark this answers is not in the document]
Comment (HenrikJannsen) [reply bisqapps/shared/domain/src/commonMain/kotlin/network/bisq/mobile/client/user_profile/ClientUserProfileServiceFacade.kt:24]: I saw you did it already. ;-)

File: bisqapps/shared/domain/src/commonMain/kotlin/network/bisq/mobile/client/service/ApiRequestService.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+import co.touchlab.kermit.Logger
+import io.ktor.client.HttpClient
+import io.ktor.client.call.body
+import io.ktor.client.request.get
+import io.ktor.client.request.post
+import io.ktor.client.request.setBody
+import io.ktor.http.contentType
+
+class ApiRequestService(val httpClient: HttpClient, host: String) {
```
Comment (rodvar) [inline bisqapps/shared/domain/src/commonMain/kotlin/network/bisq/mobile/client/service/ApiRequestService.kt:11]: :heart:

**Review by rodvar — APPROVED** [review]
:tada:

**Pull request merged**